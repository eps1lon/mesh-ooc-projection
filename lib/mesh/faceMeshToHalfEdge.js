"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const HalfEdge_1 = require("./half_edge/HalfEdge");
const DoubleLinkedList_1 = require("./half_edge/util/DoubleLinkedList");
// TODO: validate that prev,next,inverse are set
function facesToHalfEdges(mesh) {
    const edges = [];
    const faces = [];
    const vertices = mesh.vertices.map((vertex, id) => ({
        id,
        p: vertex.p,
        attributes: vertex.attributes
    }));
    let edge_id = 0;
    mesh.faces.forEach((mesh_face, face_id) => {
        const face = {
            id: face_id
        };
        const half_edges = [];
        // build half edges
        mesh_face.vertices.forEach((cur_vertex_id, i) => {
            const next_vertex_index = (i + 1) % mesh_face.vertices.length;
            const next_vertex_id = mesh_face.vertices[next_vertex_index];
            const next_vertex = vertices[next_vertex_id];
            const half_edge = {
                id: edge_id,
                target: next_vertex,
                face
            };
            // first generated edge gets referenced in face
            if (face.edge === undefined) {
                face.edge = half_edge;
            }
            // vertex doesnt have an edge yet
            if (vertices[cur_vertex_id].edge === undefined) {
                vertices[cur_vertex_id].edge = half_edge;
            }
            ++edge_id;
            half_edges.push(half_edge);
        });
        faces.push(face);
        edges.push(...DoubleLinkedList_1.doubleLink(half_edges));
    });
    const { non_manifold } = inverseMatching(edges);
    if (non_manifold > 0) {
        // tslint:disable-next-line: no-console
        console.warn('encountered non manifold edges, pathfinding might not work');
    }
    // mark every vertex with boundary edges as incomplete (see #9)
    for (const vertex of vertices) {
        vertex.attributes.boundary = HalfEdge_1.outgoingEdges(vertex).some(edge => edge.inverse === undefined);
    }
    // at this point every vertex is initialized
    return { edges, faces, vertices };
}
exports.default = facesToHalfEdges;
function inverseMatching(half_edges) {
    let borders = 0;
    let matches = 0;
    let non_manifold = 0;
    const sorted = half_edges
        .map(edge => {
        return [HalfEdge_1.origin(edge).id, edge.target.id, edge];
    })
        .map(([source, target, edge]) => {
        // internal index sort
        if (source > target) {
            return [target, source, edge];
        }
        else {
            return [source, target, edge];
        }
    })
        .sort((a, b) => {
        const primary = a[0] - b[0];
        const secondary = a[1] - b[1];
        return primary !== 0 ? primary : secondary;
    });
    for (let i = 0; i < sorted.length - 1;) {
        const edge = sorted[i][2];
        let j = i + 1;
        for (j; j < sorted.length; ++j) {
            if (sorted[i][0] !== sorted[j][0] || sorted[i][1] !== sorted[j][1]) {
                break;
            }
        }
        const pairs = j - i;
        if (pairs === 1) {
            edge.boundary = true;
            edge.manifold = true;
            ++borders;
        }
        else if (pairs === 2) {
            const inverse = sorted[i + 1][2];
            edge.inverse = inverse;
            edge.manifold = true;
            edge.boundary = false;
            inverse.inverse = edge;
            inverse.manifold = false;
            inverse.boundary = false;
            matches = matches + 2;
        }
        else {
            for (let k = i; k < j; ++k) {
                sorted[k][2].manifold = false;
                sorted[k][2].boundary = false;
            }
            non_manifold += pairs;
        }
        i = i + pairs;
    }
    return { matches, borders, non_manifold };
}
