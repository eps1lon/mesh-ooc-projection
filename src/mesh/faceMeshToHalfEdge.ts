import FaceMesh, { VertexId } from './FaceMesh';
import {
  HalfEdgeMesh,
  HalfEdge,
  Face,
  Vertex,
  outgoingEdges,
  origin
} from './half_edge/HalfEdge';
import { doubleLink } from './half_edge/util/DoubleLinkedList';

// TODO: validate that prev,next,inverse are set
export default function facesToHalfEdges(mesh: FaceMesh): HalfEdgeMesh {
  const edges: HalfEdge[] = [];
  const faces: Face[] = [];
  const vertices: Vertex[] = mesh.vertices.map(
    (vertex, id) =>
      ({
        id,
        p: vertex.p,
        attributes: vertex.attributes
      } as Vertex)
  );

  let edge_id = 0;
  mesh.faces.forEach((mesh_face, face_id) => {
    const face: Face = {
      id: face_id
    } as Face;

    const half_edges: HalfEdge[] = [];

    // build half edges
    mesh_face.vertices.forEach((cur_vertex_id, i) => {
      const next_vertex_index = (i + 1) % mesh_face.vertices.length;
      const next_vertex_id = mesh_face.vertices[next_vertex_index];
      const next_vertex = vertices[next_vertex_id];

      const half_edge: HalfEdge = {
        id: edge_id,
        target: next_vertex,
        face
      } as HalfEdge;

      // first generated edge gets referenced in face
      if (face.edge === undefined) {
        face.edge = half_edge;
      }

      // vertex doesnt have an edge yet
      if (vertices[cur_vertex_id].edge === undefined) {
        vertices[cur_vertex_id].edge = half_edge;
      }

      ++edge_id;

      half_edges.push(half_edge);
    });

    faces.push(face as Face);

    edges.push(...doubleLink(half_edges));
  });

  const { non_manifold } = inverseMatching(edges);
  if (non_manifold > 0) {
    // tslint:disable-next-line: no-console
    console.warn('encountered non manifold edges, pathfinding might not work');
  }

  // mark every vertex with boundary edges as incomplete (see #9)
  for (const vertex of vertices) {
    vertex.attributes.boundary = outgoingEdges(vertex).some(
      edge => edge.inverse === undefined
    );
  }
  // at this point every vertex is initialized
  return { edges, faces, vertices };
}

function inverseMatching(half_edges: HalfEdge[]) {
  let borders = 0;
  let matches = 0;
  let non_manifold = 0;

  const sorted = half_edges
    .map(edge => {
      return [origin(edge).id, edge.target.id, edge] as [
        VertexId,
        VertexId,
        HalfEdge
      ];
    })
    .map(([source, target, edge]) => {
      // internal index sort
      if (source > target) {
        return [target, source, edge] as [VertexId, VertexId, HalfEdge];
      } else {
        return [source, target, edge] as [VertexId, VertexId, HalfEdge];
      }
    })
    .sort((a, b) => {
      const primary = a[0] - b[0];
      const secondary = a[1] - b[1];

      return primary !== 0 ? primary : secondary;
    });

  for (let i = 0; i < sorted.length - 1; ) {
    const edge: HalfEdge = sorted[i][2];

    let j = i + 1;
    for (j; j < sorted.length; ++j) {
      if (sorted[i][0] !== sorted[j][0] || sorted[i][1] !== sorted[j][1]) {
        break;
      }
    }

    const pairs = j - i;

    if (pairs === 1) {
      edge.boundary = true;
      edge.manifold = true;

      ++borders;
    } else if (pairs === 2) {
      const inverse: HalfEdge = sorted[i + 1][2];

      edge.inverse = inverse;
      edge.manifold = true;
      edge.boundary = false;

      inverse.inverse = edge;
      inverse.manifold = false;
      inverse.boundary = false;

      matches = matches + 2;
    } else {
      for (let k = i; k < j; ++k) {
        sorted[k][2].manifold = false;
        sorted[k][2].boundary = false;
      }

      non_manifold += pairs;
    }

    i = i + pairs;
  }

  return { matches, borders, non_manifold };
}
