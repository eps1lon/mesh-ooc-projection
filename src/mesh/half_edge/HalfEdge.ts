import computeBboxBase from '../../math/bbox';
import vector, {
  add,
  dot,
  cross,
  sub,
  normalize,
  distance as vectorDistance
} from '../../math/vector';
import { normalRight } from '../../math/facette';
import {
  Normal,
  DenormalizedNormal,
  VertexAttributes as BaseVertexAttributes
} from '../FaceMesh';

// https://www.openmesh.org/Daily-Builds/Doc/a00016.html
// http://www.tandfonline.com/doi/abs/10.1080/10867651.1998.10487494
export type VertexId = number; // Uint32
export interface VertexAttributes extends BaseVertexAttributes {
  projected_heuristic: number;
  boundary: boolean;
}
export interface Vertex {
  attributes: VertexAttributes;
  id: VertexId;
  p: Vector;
  edge?: HalfEdge; // outgoing
}

export type FaceId = number; // Uint32
export interface Face {
  id: FaceId;
  edge: HalfEdge;
}

export type HalfEdgeId = number; // Uint32
export interface HalfEdge {
  id: HalfEdgeId;
  target: Vertex;
  face: Face; // face left of it
  next: HalfEdge;
  inverse?: HalfEdge;
  prev: HalfEdge;
  boundary: boolean;
  manifold: boolean;
}

export interface HalfEdgeMesh {
  edges: HalfEdge[];
  faces: Face[];
  vertices: Vertex[];
}

export function clockwise(edge: HalfEdge) {
  return edge.inverse !== undefined ? edge.inverse.next : undefined;
}

export function counterClockwise(edge: HalfEdge) {
  return edge.prev.inverse;
}

export function* circulate(
  vertex: Vertex,
  next: (edge: HalfEdge) => HalfEdge | undefined
) {
  const first_edge = vertex.edge;

  if (first_edge === undefined) {
    return;
  }

  let next_edge = next(first_edge);

  yield first_edge;

  while (next_edge !== undefined && next_edge !== first_edge) {
    yield next_edge;

    next_edge = next(next_edge);
  }
}

/**
 *
 * @param vertex
 * @return HalfEdge[] orderd counter clockwise
 */
export function outgoingEdges(vertex: Vertex): HalfEdge[] {
  // first iterate to left
  const edges_left = [...circulate(vertex, counterClockwise)]; // this.outgoindEdges(vertex, edge => edge.prev.inverse);
  // then right
  // small overhead at cost of readability => blindly iterate again
  const edges_right = [...circulate(vertex, clockwise)];

  // 1 offset for middle
  if (edges_left[edges_left.length - 1] === edges_right[1]) {
    // full circle => edges_left === edges_right
    return edges_left;
  } else {
    // counter clockwise
    return edges_left.concat(edges_right.slice(1).reverse());
  }
}

export function origin(edge: HalfEdge) {
  return edge.prev.target;
}

/**
 * Finds twin/opposite/inverse half edge of {edge}
 * O(n) where n is the number of outgoing edges from {edge.target}
 *
 * @param source
 * @param edge outgoing edge from {source}
 */
export function inverseEdge(
  source: Vertex,
  edge: HalfEdge
): HalfEdge | undefined {
  return outgoingEdges(edge.target).find(other => other.target === source);
}

/**
 * @param triangle
 * @return 3d coords of each vertex in {face}
 */
export function facePoints(triangle: Face): [Vector, Vector, Vector] {
  const u = triangle.edge;
  const v = u.next;
  const w = v.next;

  return [u.target.p, v.target.p, w.target.p];
}

export function computeBbox(mesh: HalfEdgeMesh) {
  return computeBboxBase(mesh.vertices, vertex => vertex.p);
}

/**
 * iterates over each edge that shares a face with {start}
 * excludes {start}
 * @param start
 */
export function* faceEdges(start: HalfEdge) {
  let cur = start.next;

  while (cur !== start) {
    yield cur;
    cur = cur.next;
  }
}

export function distance(source: Vertex, target: Vertex) {
  return vectorDistance(source.p, target.p);
}

export function innerAngle(e1: HalfEdge, e2: HalfEdge): number {
  const N = plane(e1, e2);

  const p1 = origin(e1).p;
  const p2 = e1.target.p;
  const p3 = e2.target.p;

  const u = normalize(sub(p1, p2));
  const v = normalize(sub(p3, p2));

  const cosine = dot(u, v);
  let angle = Math.acos(cosine);

  // rounding error
  if (Number.isNaN(angle)) {
    if (cosine > 0) {
      angle = Number.EPSILON;
    } else {
      angle = Math.PI - Number.EPSILON;
    }
  }

  if (dot(cross(u, v), N) < 0) {
    return 2 * Math.PI - angle;
  } else {
    return angle;
  }
}

export function plane(...edges: HalfEdge[]): Normal {
  let n: DenormalizedNormal = vector(0, 0, 0);

  for (const edge of edges) {
    n = add(n, normalRight(facePoints(edge.face)));
    if (edge.inverse !== undefined) {
      n = add(n, normalRight(facePoints(edge.inverse.face)));
    }
  }

  return normalize(n);
}
