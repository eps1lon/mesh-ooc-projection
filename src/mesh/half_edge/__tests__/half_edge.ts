/// <reference types="@types/jest" />
import { facePoints } from '../HalfEdge';
import faceMeshToHalfEdge from '../../faceMeshToHalfEdge';
import big_mesh from './__fixtures__/mesh';

it('should dereference face vertices', () => {
  const mesh = faceMeshToHalfEdge(big_mesh);

  expect(mesh.faces.map(face => facePoints(face))).toMatchSnapshot();
});
