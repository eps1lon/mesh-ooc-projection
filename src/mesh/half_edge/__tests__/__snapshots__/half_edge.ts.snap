// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should dereference face vertices 1`] = `
Array [
  Array [
    Array [
      -20.06145,
      -301.5033,
      107.4102,
    ],
    Array [
      -19.64998,
      -301.6511,
      107.5756,
    ],
    Array [
      -19.87812,
      -301.7766,
      107.7385,
    ],
  ],
  Array [
    Array [
      -18.99838,
      -302.3419,
      107.6257,
    ],
    Array [
      -18.29505,
      -302.4955,
      107.7116,
    ],
    Array [
      -18.45564,
      -302.4262,
      107.5372,
    ],
  ],
  Array [
    Array [
      -18.5668,
      -302.6424,
      107.4552,
    ],
    Array [
      -19.52846,
      -302.3832,
      107.6959,
    ],
    Array [
      -18.99838,
      -302.3419,
      107.6257,
    ],
  ],
  Array [
    Array [
      -18.99838,
      -302.3419,
      107.6257,
    ],
    Array [
      -18.45564,
      -302.4262,
      107.5372,
    ],
    Array [
      -18.5668,
      -302.6424,
      107.4552,
    ],
  ],
  Array [
    Array [
      -18.5668,
      -302.6424,
      107.4552,
    ],
    Array [
      -18.96306,
      -302.7753,
      107.5695,
    ],
    Array [
      -19.52846,
      -302.3832,
      107.6959,
    ],
  ],
  Array [
    Array [
      -19.52846,
      -302.3832,
      107.6959,
    ],
    Array [
      -18.96306,
      -302.7753,
      107.5695,
    ],
    Array [
      -19.33993,
      -302.8016,
      107.6754,
    ],
  ],
  Array [
    Array [
      -18.29505,
      -302.4955,
      107.7116,
    ],
    Array [
      -18.5668,
      -302.6424,
      107.4552,
    ],
    Array [
      -18.45564,
      -302.4262,
      107.5372,
    ],
  ],
  Array [
    Array [
      -18.96306,
      -302.7753,
      107.5695,
    ],
    Array [
      -18.5668,
      -302.6424,
      107.4552,
    ],
    Array [
      -18.56424,
      -302.8596,
      107.6676,
    ],
  ],
  Array [
    Array [
      -18.56424,
      -302.8596,
      107.6676,
    ],
    Array [
      -18.5668,
      -302.6424,
      107.4552,
    ],
    Array [
      -18.29505,
      -302.4955,
      107.7116,
    ],
  ],
  Array [
    Array [
      -18.2355,
      -304.7742,
      107.5723,
    ],
    Array [
      -18.61397,
      -304.9963,
      107.6061,
    ],
    Array [
      -18.61863,
      -304.7247,
      107.7419,
    ],
  ],
  Array [
    Array [
      -18.2355,
      -304.7742,
      107.5723,
    ],
    Array [
      -18.33511,
      -305.1866,
      107.5966,
    ],
    Array [
      -18.61397,
      -304.9963,
      107.6061,
    ],
  ],
  Array [
    Array [
      -18.08405,
      -304.9866,
      107.7483,
    ],
    Array [
      -18.33511,
      -305.1866,
      107.5966,
    ],
    Array [
      -18.2355,
      -304.7742,
      107.5723,
    ],
  ],
  Array [
    Array [
      -18.61397,
      -304.9963,
      107.6061,
    ],
    Array [
      -18.33511,
      -305.1866,
      107.5966,
    ],
    Array [
      -18.59071,
      -305.4381,
      107.7289,
    ],
  ],
  Array [
    Array [
      -18.59071,
      -305.4381,
      107.7289,
    ],
    Array [
      -18.37235,
      -305.7494,
      107.5651,
    ],
    Array [
      -18.66739,
      -305.6412,
      107.6331,
    ],
  ],
  Array [
    Array [
      -22.11425,
      -306.645,
      107.722,
    ],
    Array [
      -21.84192,
      -306.8357,
      107.5388,
    ],
    Array [
      -22.10112,
      -306.9336,
      107.7118,
    ],
  ],
  Array [
    Array [
      -21.70459,
      -306.7019,
      107.6875,
    ],
    Array [
      -21.84192,
      -306.8357,
      107.5388,
    ],
    Array [
      -22.11425,
      -306.645,
      107.722,
    ],
  ],
  Array [
    Array [
      -18.66739,
      -305.6412,
      107.6331,
    ],
    Array [
      -18.46808,
      -306.1833,
      107.4703,
    ],
    Array [
      -18.81029,
      -306.4738,
      107.5788,
    ],
  ],
  Array [
    Array [
      -18.66739,
      -305.6412,
      107.6331,
    ],
    Array [
      -18.37235,
      -305.7494,
      107.5651,
    ],
    Array [
      -18.46808,
      -306.1833,
      107.4703,
    ],
  ],
  Array [
    Array [
      -18.46808,
      -306.1833,
      107.4703,
    ],
    Array [
      -18.37235,
      -305.7494,
      107.5651,
    ],
    Array [
      -18.16206,
      -306.1879,
      107.5233,
    ],
  ],
  Array [
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
    Array [
      -23.64511,
      -308.568,
      107.6906,
    ],
    Array [
      -23.38466,
      -308.4229,
      107.667,
    ],
  ],
  Array [
    Array [
      -23.38466,
      -308.4229,
      107.667,
    ],
    Array [
      -22.57956,
      -308.6053,
      107.562,
    ],
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
  ],
  Array [
    Array [
      -22.4844,
      -309.1028,
      107.5584,
    ],
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
    Array [
      -22.57956,
      -308.6053,
      107.562,
    ],
  ],
  Array [
    Array [
      -21.71572,
      -308.5911,
      107.6888,
    ],
    Array [
      -21.92677,
      -308.9722,
      107.6574,
    ],
    Array [
      -22.57956,
      -308.6053,
      107.562,
    ],
  ],
  Array [
    Array [
      -22.4844,
      -309.1028,
      107.5584,
    ],
    Array [
      -22.57956,
      -308.6053,
      107.562,
    ],
    Array [
      -21.92677,
      -308.9722,
      107.6574,
    ],
  ],
  Array [
    Array [
      -21.55886,
      -308.9018,
      107.5922,
    ],
    Array [
      -21.92677,
      -308.9722,
      107.6574,
    ],
    Array [
      -21.71572,
      -308.5911,
      107.6888,
    ],
  ],
  Array [
    Array [
      -23.64511,
      -308.568,
      107.6906,
    ],
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
    Array [
      -23.50346,
      -309.5736,
      107.6632,
    ],
  ],
  Array [
    Array [
      -22.4844,
      -309.1028,
      107.5584,
    ],
    Array [
      -21.92677,
      -308.9722,
      107.6574,
    ],
    Array [
      -22.15418,
      -309.2999,
      107.5305,
    ],
  ],
  Array [
    Array [
      -21.55886,
      -308.9018,
      107.5922,
    ],
    Array [
      -21.73598,
      -309.3085,
      107.6865,
    ],
    Array [
      -21.92677,
      -308.9722,
      107.6574,
    ],
  ],
  Array [
    Array [
      -21.73598,
      -309.3085,
      107.6865,
    ],
    Array [
      -22.15418,
      -309.2999,
      107.5305,
    ],
    Array [
      -21.92677,
      -308.9722,
      107.6574,
    ],
  ],
  Array [
    Array [
      -21.55886,
      -308.9018,
      107.5922,
    ],
    Array [
      -21.40649,
      -309.0595,
      107.7266,
    ],
    Array [
      -21.73598,
      -309.3085,
      107.6865,
    ],
  ],
  Array [
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
    Array [
      -23.10824,
      -309.3931,
      107.5181,
    ],
    Array [
      -23.50346,
      -309.5736,
      107.6632,
    ],
  ],
  Array [
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
    Array [
      -23.10824,
      -309.3931,
      107.5181,
    ],
  ],
  Array [
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
    Array [
      -22.9475,
      -308.9085,
      107.5073,
    ],
    Array [
      -22.4844,
      -309.1028,
      107.5584,
    ],
  ],
  Array [
    Array [
      -22.15418,
      -309.2999,
      107.5305,
    ],
    Array [
      -22.32691,
      -309.7278,
      107.5688,
    ],
    Array [
      -22.4844,
      -309.1028,
      107.5584,
    ],
  ],
  Array [
    Array [
      -21.73598,
      -309.3085,
      107.6865,
    ],
    Array [
      -22.32691,
      -309.7278,
      107.5688,
    ],
    Array [
      -22.15418,
      -309.2999,
      107.5305,
    ],
  ],
  Array [
    Array [
      -23.75085,
      -310.1497,
      107.7151,
    ],
    Array [
      -23.50346,
      -309.5736,
      107.6632,
    ],
    Array [
      -23.10824,
      -309.3931,
      107.5181,
    ],
  ],
  Array [
    Array [
      -22.4844,
      -309.1028,
      107.5584,
    ],
    Array [
      -22.32691,
      -309.7278,
      107.5688,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
  ],
  Array [
    Array [
      -21.73598,
      -309.3085,
      107.6865,
    ],
    Array [
      -21.77079,
      -309.7588,
      107.5304,
    ],
    Array [
      -22.32691,
      -309.7278,
      107.5688,
    ],
  ],
  Array [
    Array [
      -23.10824,
      -309.3931,
      107.5181,
    ],
    Array [
      -23.08648,
      -310.0202,
      107.5497,
    ],
    Array [
      -23.75085,
      -310.1497,
      107.7151,
    ],
  ],
  Array [
    Array [
      -23.08648,
      -310.0202,
      107.5497,
    ],
    Array [
      -23.10824,
      -309.3931,
      107.5181,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
  ],
  Array [
    Array [
      -23.08648,
      -310.0202,
      107.5497,
    ],
    Array [
      -23.37333,
      -310.4771,
      107.7121,
    ],
    Array [
      -23.75085,
      -310.1497,
      107.7151,
    ],
  ],
  Array [
    Array [
      -23.37333,
      -310.4771,
      107.7121,
    ],
    Array [
      -23.08648,
      -310.0202,
      107.5497,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
  ],
  Array [
    Array [
      -23.37333,
      -310.4771,
      107.7121,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
    Array [
      -22.92031,
      -310.7228,
      107.5577,
    ],
  ],
  Array [
    Array [
      -22.92031,
      -310.7228,
      107.5577,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
  ],
  Array [
    Array [
      -21.77079,
      -309.7588,
      107.5304,
    ],
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
    Array [
      -22.32691,
      -309.7278,
      107.5688,
    ],
  ],
  Array [
    Array [
      -22.32691,
      -309.7278,
      107.5688,
    ],
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
    Array [
      -22.70528,
      -309.7769,
      107.4795,
    ],
  ],
  Array [
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
    Array [
      -21.77079,
      -309.7588,
      107.5304,
    ],
    Array [
      -21.77134,
      -310.19,
      107.5216,
    ],
  ],
  Array [
    Array [
      -21.33407,
      -310.0416,
      107.6961,
    ],
    Array [
      -21.77134,
      -310.19,
      107.5216,
    ],
    Array [
      -21.77079,
      -309.7588,
      107.5304,
    ],
  ],
  Array [
    Array [
      -22.92031,
      -310.7228,
      107.5577,
    ],
    Array [
      -23.30143,
      -310.8515,
      107.6351,
    ],
    Array [
      -23.37333,
      -310.4771,
      107.7121,
    ],
  ],
  Array [
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
    Array [
      -21.77134,
      -310.19,
      107.5216,
    ],
    Array [
      -22.13051,
      -310.4781,
      107.639,
    ],
  ],
  Array [
    Array [
      -21.33407,
      -310.0416,
      107.6961,
    ],
    Array [
      -21.50583,
      -310.4327,
      107.6641,
    ],
    Array [
      -21.77134,
      -310.19,
      107.5216,
    ],
  ],
  Array [
    Array [
      -22.13051,
      -310.4781,
      107.639,
    ],
    Array [
      -21.77134,
      -310.19,
      107.5216,
    ],
    Array [
      -21.50583,
      -310.4327,
      107.6641,
    ],
  ],
  Array [
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
    Array [
      -22.6717,
      -310.8226,
      107.5072,
    ],
    Array [
      -22.92031,
      -310.7228,
      107.5577,
    ],
  ],
  Array [
    Array [
      -22.46815,
      -310.3087,
      107.5531,
    ],
    Array [
      -22.13051,
      -310.4781,
      107.639,
    ],
    Array [
      -22.6717,
      -310.8226,
      107.5072,
    ],
  ],
  Array [
    Array [
      -21.50583,
      -310.4327,
      107.6641,
    ],
    Array [
      -21.68603,
      -310.9463,
      107.7246,
    ],
    Array [
      -22.13051,
      -310.4781,
      107.639,
    ],
  ],
  Array [
    Array [
      -22.92031,
      -310.7228,
      107.5577,
    ],
    Array [
      -22.6717,
      -310.8226,
      107.5072,
    ],
    Array [
      -23.30143,
      -310.8515,
      107.6351,
    ],
  ],
  Array [
    Array [
      -22.13051,
      -310.4781,
      107.639,
    ],
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
    Array [
      -22.6717,
      -310.8226,
      107.5072,
    ],
  ],
  Array [
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
    Array [
      -23.3157,
      -311.209,
      107.6803,
    ],
    Array [
      -23.30143,
      -310.8515,
      107.6351,
    ],
  ],
  Array [
    Array [
      -23.30143,
      -310.8515,
      107.6351,
    ],
    Array [
      -22.6717,
      -310.8226,
      107.5072,
    ],
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
  ],
  Array [
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
    Array [
      -22.13051,
      -310.4781,
      107.639,
    ],
    Array [
      -21.68603,
      -310.9463,
      107.7246,
    ],
  ],
  Array [
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
    Array [
      -22.97901,
      -311.5553,
      107.6225,
    ],
    Array [
      -23.3157,
      -311.209,
      107.6803,
    ],
  ],
  Array [
    Array [
      -21.68603,
      -310.9463,
      107.7246,
    ],
    Array [
      -22.36991,
      -311.5438,
      107.6351,
    ],
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
  ],
  Array [
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
    Array [
      -22.73971,
      -311.8039,
      107.6383,
    ],
    Array [
      -22.97901,
      -311.5553,
      107.6225,
    ],
  ],
  Array [
    Array [
      -23.4891,
      -312.2336,
      107.6915,
    ],
    Array [
      -22.97901,
      -311.5553,
      107.6225,
    ],
    Array [
      -22.73971,
      -311.8039,
      107.6383,
    ],
  ],
  Array [
    Array [
      -22.73971,
      -311.8039,
      107.6383,
    ],
    Array [
      -22.36991,
      -311.5438,
      107.6351,
    ],
    Array [
      -22.45631,
      -312.152,
      107.6031,
    ],
  ],
  Array [
    Array [
      -22.73971,
      -311.8039,
      107.6383,
    ],
    Array [
      -22.55308,
      -311.33,
      107.6484,
    ],
    Array [
      -22.36991,
      -311.5438,
      107.6351,
    ],
  ],
  Array [
    Array [
      -22.00361,
      -312.4008,
      107.6966,
    ],
    Array [
      -22.45631,
      -312.152,
      107.6031,
    ],
    Array [
      -22.36991,
      -311.5438,
      107.6351,
    ],
  ],
  Array [
    Array [
      -22.73971,
      -311.8039,
      107.6383,
    ],
    Array [
      -23.03733,
      -312.4767,
      107.6988,
    ],
    Array [
      -23.4891,
      -312.2336,
      107.6915,
    ],
  ],
  Array [
    Array [
      -23.03733,
      -312.4767,
      107.6988,
    ],
    Array [
      -22.73971,
      -311.8039,
      107.6383,
    ],
    Array [
      -22.45631,
      -312.152,
      107.6031,
    ],
  ],
  Array [
    Array [
      -22.00361,
      -312.4008,
      107.6966,
    ],
    Array [
      -22.45512,
      -312.97,
      107.7332,
    ],
    Array [
      -22.45631,
      -312.152,
      107.6031,
    ],
  ],
  Array [
    Array [
      -22.00361,
      -312.4008,
      107.6966,
    ],
    Array [
      -22.03425,
      -312.9892,
      107.6969,
    ],
    Array [
      -22.45512,
      -312.97,
      107.7332,
    ],
  ],
]
`;
